### ISSUE TITLE
Batch of Multinomials - Error

### ISSUE THREAD

**Issue by tudaga** [blei-lab/edward#748]
When I try to create a batch of Multinomials by following the example in [http://edwardlib.org/api/ed/models/Multinomial](http://edwardlib.org/api/ed/models/Multinomial):

```
p = [[.1, .2, .7], [.3, .3, .4]]  # Shape [2, 3]
dist = Multinomial(total_count=[4., 5], probs=p)
```

I get the following error message:
```
NotImplementedError: sample is not implemented for Multinomial. 
You must either pass in the value argument or implement sample for Multinomial.
```

**Comment by dustinvtran** [comment]
It looks like batches aren't currently supported by TensorFlow doesn't support non-scalar multinomial draws: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/distributions/multinomial.py#L236

**Comment by tudaga** [comment]
Thanks for the quick reply! This is a little bit confusing because I just made the code snippet work in tensorflow 1.2 (also in 1.3)
```
from tensorflow.contrib.distributions import Multinomial
```
but it doesn't work in edward :(

**Comment by dustinvtran** [comment]
Can you show your working snippet? `_sample_n` for Multinomial hasn't changed in r1.2 so not sure how it could work in either if `if self.total_count.get_shape().ndims != 0`.

**Comment by tudaga** [comment]
Here is an example of a working snippet:
```
from tensorflow.contrib.distributions import Multinomial
p = [[.1, .2, .7], [.3, .3, .4]]  # Shape [2, 3]
dist = Multinomial(total_count=[4., 5], probs=p)
counts = [[2., 1, 1], [3, 1, 1]]
sess = tf.Session()
with sess.as_default():
    print(dist.prob(counts).eval())
```
Thanks again for the fast reply!

**Comment by dustinvtran** [comment]
Random variables in Edward are light wrappers around TensorFlow distributions. When instantiating Edward random variables, we call sample on them which would still cause an error here. More details in the model API (http://edwardlib.org/api/model).

**Comment by tudaga** [comment]
Thanks a lot for the explanation, I understand the issue now. I have submitted a feature request to tensorflow.

**Comment by tudaga** [comment]
I am using the Multinomial distribution in Edward in the context of variational inference. I can see how drawing from it will be necessary for exact inference or MCMC, but it seems like sampling wouldn't be necessary for variational inference. Is there a possible way to avoid the issue in my case?

**Comment by dustinvtran** [comment]
You can avoid sampling if it's only used for the likelihood. See the last suggestion in the custom random variable API (http://edwardlib.org/api/model-development).

**Comment by tudaga** [comment]
Thanks, I see. Unfortunately it is not in the likelihood. The model I'm trying to implement is hierarchical poisson factorization [http://www.cs.columbia.edu/~blei/papers/GopalanHofmanBlei2015.pdf](http://www.cs.columbia.edu/~blei/papers/GopalanHofmanBlei2015.pdf)
I was wrong before when I said I do not need to sample from it. It seems that in the Edward KLqp some form of black box VI is always used (even for conditionally conjugate models like in my case), and therefore sampling from the Multinomial is necessary.

**Comment by dawenl** [comment]
For conditional conjugate models (like HPF), it's recommended to do closed-form coordinate updates over SGD based black-box VI -- it's very possible you will run into convergence issue with realistically-sized data.

**Comment by tudaga** [comment]
Thanks, this is very useful information. I'll move on to trying to implement the closed-form coordinate updates! Do you recommend that I do this in tensorflow @dawenl ?

**Comment by dawenl** [comment]
I don't think it would be very helpful with tensorflow since you cannot make use of automatic differentiation. Standard numpy/scipy should be enough.

**Comment by dawenl** [comment]
This could actually be a challenging case for the conjugacy algebra :) I am curious to see if it can work out this model. You probably need to add the necessary auxiliary random variables for Poisson factorization to make the conjugacy algebra work.
